File: run-migrations-via-api.ts
#!/usr/bin/env bun
/**
 * Run fairy tale migrations via API
 * Reads SQL files and sends them to the backend API endpoint
 */

import { readFile } from "fs/promises";
import { join } from "path";

// Backend API URL
const BACKEND_URL = "https://backend-2-production-3de1.up.railway.app";
const API_ENDPOINT = `${BACKEND_URL}/fairytales/run-migration-sql`;

async function runMigration(migrationPath: string, migrationName: string): Promise<boolean> {
  console.log(`\n🔄 Running ${migrationName}...`);

  try {
    // Read SQL file
    const sql = await readFile(migrationPath, "utf-8");
    console.log(`  📄 SQL file size: ${sql.length} characters`);

    // Send to API
    const response = await fetch(API_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        sql: sql,
        migrationName: migrationName,
      }),
    });

    if (response.ok) {
      const result = await response.json();
      if (result.success) {
        console.log(`  ✅ ${migrationName} completed successfully`);
        console.log(`     ${result.message || ""}`);
        return true;
      } else {
        console.log(`  ❌ ${migrationName} failed`);
        console.log(`     ${result.message || ""}`);
        return false;
      }
    } else {
      console.log(`  ❌ HTTP Error ${response.status}`);
      const text = await response.text();
      console.log(`     ${text}`);
      return false;
    }
  } catch (error: any) {
    console.log(`  ❌ Error: ${error.message}`);
    return false;
  }
}

async function main() {
  console.log("🚀 Talea Fairy Tales Migration Runner (API Mode)\n");

  // Define migrations in order
  const migrationsDir = join(import.meta.dir, "backend", "fairytales", "migrations");
  const migrations = [
    { file: "10_add_47_classic_fairy_tales.up.sql", name: "10_add_47_classic_fairy_tales" },
    { file: "11_add_andersen_fairy_tales.up.sql", name: "11_add_andersen_fairy_tales" },
    { file: "12_add_russian_arabian_fairy_tales.up.sql", name: "12_add_russian_arabian_fairy_tales" },
    { file: "13_add_classics_legends_fables.up.sql", name: "13_add_classics_legends_fables" },
  ];

  // Check current count
  console.log("📊 Checking current fairy tale count...");
  try {
    const response = await fetch(`${BACKEND_URL}/fairytales/trigger-migrations`);
    if (response.ok) {
      const data = await response.json();
      const currentCount = data.taleCount || 0;
      console.log(`  Current count: ${currentCount} tales\n`);

      if (currentCount >= 50) {
        console.log("✅ Database already has 50+ fairy tales. No migrations needed.");
        return;
      }
    } else {
      console.log(`  ⚠️  Could not check current count (continuing anyway)`);
    }
  } catch (error: any) {
    console.log(`  ⚠️  Error checking count: ${error.message} (continuing anyway)\n`);
  }

  // Run each migration
  let successCount = 0;
  for (const migration of migrations) {
    const migrationPath = join(migrationsDir, migration.file);

    const success = await runMigration(migrationPath, migration.name);
    if (success) {
      successCount++;
    } else {
      console.log(`\n⚠️  Migration failed. Stopping here.`);
      break;
    }
  }

  // Check final count
  console.log(`\n📊 Final Results:`);
  console.log(`  Migrations executed: ${successCount}/${migrations.length}`);

  try {
    const response = await fetch(`${BACKEND_URL}/fairytales/trigger-migrations`);
    if (response.ok) {
      const data = await response.json();
      const finalCount = data.taleCount || 0;
      console.log(`  Final fairy tale count: ${finalCount}`);

      if (finalCount === 50) {
        console.log("\n🎉 SUCCESS! Database now has exactly 50 fairy tales!");
        console.log("\n🌐 Check them out at: https://www.talea.website/fairytales");
      } else if (finalCount > 50) {
        console.log(`\n⚠️  Warning: Database has ${finalCount} tales (expected 50)`);
      } else {
        console.log(`\n⚠️  Warning: Only ${finalCount} tales found (expected 50)`);
      }
    }
  } catch (error: any) {
    console.log(`\n⚠️  Could not check final count: ${error.message}`);
  }
}

main().catch(console.error);

File: run-migrations-via-api.py
#!/usr/bin/env python3
"""
Run fairy tale migrations via API
Reads SQL files and sends them to the backend API endpoint
"""

import requests
import sys
from pathlib import Path

# Backend API URL
BACKEND_URL = "https://backend-2-production-3de1.up.railway.app"
API_ENDPOINT = f"{BACKEND_URL}/fairytales/run-migration-sql"

def run_migration(migration_file: Path):
    """Execute a single migration file via API"""
    print(f"\n🔄 Running {migration_file.name}...")

    try:
        # Read SQL file
        sql = migration_file.read_text(encoding='utf-8')
        print(f"  📄 SQL file size: {len(sql)} characters")

        # Send to API
        response = requests.post(
            API_ENDPOINT,
            json={
                "sql": sql,
                "migrationName": migration_file.name
            },
            timeout=120  # 2 minutes timeout for large migrations
        )

        if response.status_code == 200:
            result = response.json()
            if result.get("success"):
                print(f"  ✅ {migration_file.name} completed successfully")
                print(f"     {result.get('message', '')}")
                return True
            else:
                print(f"  ❌ {migration_file.name} failed")
                print(f"     {result.get('message', '')}")
                return False
        else:
            print(f"  ❌ HTTP Error {response.status_code}")
            print(f"     {response.text}")
            return False

    except Exception as e:
        print(f"  ❌ Error: {e}")
        return False

def main():
    print("🚀 Talea Fairy Tales Migration Runner (API Mode)\n")

    # Define migrations in order
    migrations_dir = Path(__file__).parent / "backend" / "fairytales" / "migrations"
    migrations = [
        migrations_dir / "10_add_47_classic_fairy_tales.up.sql",
        migrations_dir / "11_add_andersen_fairy_tales.up.sql",
        migrations_dir / "12_add_russian_arabian_fairy_tales.up.sql",
        migrations_dir / "13_add_classics_legends_fables.up.sql",
    ]

    # Check current count
    print("📊 Checking current fairy tale count...")
    try:
        response = requests.get(f"{BACKEND_URL}/fairytales/trigger-migrations")
        if response.status_code == 200:
            data = response.json()
            current_count = data.get("taleCount", 0)
            print(f"  Current count: {current_count} tales\n")

            if current_count >= 50:
                print("✅ Database already has 50+ fairy tales. No migrations needed.")
                return
        else:
            print(f"  ⚠️  Could not check current count (continuing anyway)")
    except Exception as e:
        print(f"  ⚠️  Error checking count: {e} (continuing anyway)\n")

    # Run each migration
    success_count = 0
    for migration_file in migrations:
        if not migration_file.exists():
            print(f"⚠️  Migration file not found: {migration_file}")
            continue

        if run_migration(migration_file):
            success_count += 1
        else:
            print(f"\n⚠️  Migration failed. Stopping here.")
            break

    # Check final count
    print(f"\n📊 Final Results:")
    print(f"  Migrations executed: {success_count}/{len(migrations)}")

    try:
        response = requests.get(f"{BACKEND_URL}/fairytales/trigger-migrations")
        if response.status_code == 200:
            data = response.json()
            final_count = data.get("taleCount", 0)
            print(f"  Final fairy tale count: {final_count}")

            if final_count == 50:
                print("\n🎉 SUCCESS! Database now has exactly 50 fairy tales!")
            elif final_count > 50:
                print(f"\n⚠️  Warning: Database has {final_count} tales (expected 50)")
            else:
                print(f"\n⚠️  Warning: Only {final_count} tales found (expected 50)")
    except Exception as e:
        print(f"\n⚠️  Could not check final count: {e}")

if __name__ == "__main__":
    main()
